src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -216,18 +216,14 @@ private async ValueTask<CancellationTokenRegistration> HandleWriteStartState(Can
                 throw new InvalidOperationException(SR.net_quic_writing_notallowed);
             }
 
-            lock (_state)
+            // Make sure start has completed
+            if (!_started)
             {
-                if (_state.SendState == SendState.Aborted)
-                {
-                    throw new OperationCanceledException(SR.net_quic_sending_aborted);
-                }
-                else if (_state.SendState == SendState.ConnectionClosed)
-                {
-                    throw GetConnectionAbortedException(_state);
-                }
+                await _state.SendResettableCompletionSource.GetTypelessValueTask().ConfigureAwait(false);
```

Comment:
If I don't wait for it before creating CancellationTokenRegistration, in case of pre-cancelled token, SendResettableCompletionSource.CompleteException fails with `InvalidOperationException: Operation is not valid due to the current state of the object.` 🤔

Reply:
Yeah, we guard the `ResettableCompletionSource` completion with `SendState`, but it "breaks" here since we're (ab)using `SendResettableCompletionSource` for `START_COMPLETE` as well.
I don't have any better suggestions than what you did here.